lib/src/Pos/Diffusion/Full/Block.hs

```diff
@@ -154,19 +156,35 @@ getBlocks
     -> [HeaderHash]
     -> d (Either GetBlocksError [Block])
 getBlocks logic enqueue nodeId tipHeader checkpoints = do
-    headers <- requestHeaders
-    blocks <- requestBlocks headers
-    return (Right (toList blocks))
-    -- TODO exception handling? Don't just catchj all.
+    -- It is apparently an error to request headers for the tipHeader and
```

Comment:
Yes, there is something super strange going here. It seems I missed the original interface planning, but `getBlocks` shouldn't be designed in this way. The communication routine "ask for checkpoints, then ask for blocks" is application logic, not diffusion layer. Diffusion layer is responsible for `getBlocks` downloading blocks concurrently from many nodes, it shouldn't know anything about the protocol except for messages imo.

Reply:
> The communication routine "ask for checkpoints, then ask for blocks" is application logic, not diffusion layer.

I'm really not so sure about that. The common case is for the application to demand some blocks. Are there legitimate common cases in which the application needs headers but not any bodies?

Reply:
Asking for tip is asking for headers, as it has the same message type. There are no other use cases of "ask for headers only" indeed, but it's not the point. Logic level benefits from having more flexible networking interface, as it performs classification between requesting headers and asking for blocks.

Reply:
> Logic level benefits from having more flexible networking interface, as it performs classification between requesting headers and asking for blocks.

Is this really a benefit? Why should the requested classify the headers? Surely the system would be more efficient if the server, which presumably has the chain that the client wants, could just send the blocks right away.

Reply:
Requester shouldn't classify headers, sure. Let's move this discussion to https://github.com/input-output-hk/cardano-sl/pull/2154#discussion_r160946710